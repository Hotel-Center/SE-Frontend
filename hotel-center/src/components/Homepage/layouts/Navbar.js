import React, {
  useState,
  useEffect,
  useCallback,
  Component,
  Fragment,
} from "react";
import Logo from "../../../statics/logo/logo2.png";
import UseAnimations from "react-useanimations";
import menu3 from "react-useanimations/lib/menu3";
import Signup from "../../Sign_up/sign_up";
import Login from "../../Login/Login";
import { me, logout } from "../../../Utils/connection";
import { cookies } from "../../../Utils/common";
import { fontFamily, fontStyle } from "@mui/system";
import KeyboardArrowDownIcon from "@mui/icons-material/KeyboardArrowDown";
import FavoriteBorderOutlinedIcon from "@mui/icons-material/FavoriteBorderOutlined";

class Navbar extends Component {
  constructor() {
    super();
    this.state = {
      openMenu: false,
      navbarMoved: false,
      is_loggedin: false,
    };
  }

  componentDidMount() {
    document.scrollingElement.scrollTop = 0;
    document.addEventListener("scroll", this.handleScroll);
    me().then((response) => {
      this.setState({ is_loggedin: response });
    });
  }

  componentWillUnmount() {
    document.removeEventListener("scroll", this.handleScroll);
  }

  handleOpenMenu = () => {
    this.state.openMenu
      ? this.setState({ openMenu: false })
      : this.setState({ openMenu: true });
  };

  handleScroll = () => {
    var scrolled = document.scrollingElement.scrollTop;
    if (scrolled >= 10) {
      if (!this.state.navbarMoved) {
        this.setState({
          navbarMoved: true,
        });
      }
    } else {
      if (this.state.navbarMoved) {
        this.setState({
          navbarMoved: false,
        });
      }
    }
  };

  handleLogout = () => {
    logout();
  };

  render() {
    return (
      <nav
        className={
          this.state.navbarMoved
            ? "navbar navbar-expand-lg navbar-light sticky-top nav-scrolled w-100 nav-style"
            : "navbar navbar-expand-lg navbar-light sticky-top nav-top w-100 nav-style"
        }
      >
        <div className="container-fluid">
          <a href="/" className="navbar-brand logo">
            <img src={Logo} alt="Hotel Center" />
            <span className="fw-bold logo-text-font">Hotel Center</span>
          </a>

          <button
            className="navbar-toggler nav-hamburger-button"
            type="button"
            data-bs-toggle="collapse"
            data-bs-target="#navMenu"
          >
            <UseAnimations
              reverse={this.state.openMenu}
              onClick={this.handleOpenMenu}
              size={40}
              animation={menu3}
              speed={3}
              strokeColor="#cd9a2d"
            />
          </button>
          <div className="collapse navbar-collapse" id="navMenu">
            <div className="ms-auto pt-3 pt-sm-0">
              <ul className="navbar-nav">
                <li className="nav-item">
                  <a
                    className="nav-link nav-menu-style"
                    aria-current="page"
                    href="#"
                  >
                    Hotels
                  </a>
                </li>
                <li className="nav-item">
                  <a className="nav-link nav-menu-style" href="#">
                    Rooms
                  </a>
                </li>
                <li className="nav-item dropdown">
                  <a
                    className="nav-link dropdown-toggle nav-menu-style"
                    href="#"
                    id="navbarDropdown"
                    role="button"
                    data-bs-toggle="dropdown"
                    aria-expanded="false"
                  >
                    Dropdown
                  </a>

                  <ul
                    className="dropdown-menu mt-2"
                    aria-labelledby="navbarDropdown"
                  >
                    <li>
                      <a className="dropdown-item" href="#">
                        Action
                      </a>
                    </li>
                    <li>
                      <a className="dropdown-item" href="#">
                        Another action
                      </a>
                    </li>
                  </ul>
                </li>
                <li className="nav-item">
                  <a className="nav-link nav-menu-style" href="#">
                    Domain
                  </a>
                </li>
                <li className="nav-item">
                  <a className="nav-link nav-menu-style" href="#">
                    Pricing
                  </a>
                </li>
                <li className="nav-item">
                  <a className="nav-link nav-menu-style" href="#">
                    Favorites
                  </a>
                </li>
                {!this.state.is_loggedin ? (
                  <Fragment>
                    <li>
                      <a href="http://localhost:3000/login">
                        <button
                          type="button"
                          className="btn btn-outline-dark me-2 nav-button nav-menu-style"
                        >
                          Login
                        </button>
                      </a>
                    </li>
                    <li>
                      <a href="http://localhost:3000/sign-up">
                        <button
                          type="button"
                          className="btn btn-outline-dark nav-button nav-menu-style"
                        >
                          Sign up
                        </button>
                      </a>
                    </li>
                  </Fragment>
                ) : (
                  <button
                    type="button"
                    className="btn btn-outline-dark me-2 nav-button fw-bold"
                    onClick={() => {
                      this.handleLogout();
                    }}
                  >
                    logout
                  </button>
                )}
              </ul>
            </div>
          </div>
        </div>
      </nav>
    );
  }
}

export default Navbar;
