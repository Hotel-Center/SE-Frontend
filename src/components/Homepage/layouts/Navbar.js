import React, { Fragment, useEffect, useReducer } from "react";
import Logo from "../../../../public/logo/logo2.png";
// import UseAnimations from "react-useanimations";
// import menu3 from "react-useanimations/lib/menu3";
import { me, logout } from "../../../Utils/connection";
import Avatar from "@mui/material/Avatar";
import PersonIcon from "@mui/icons-material/Person";
import IconButton from "@mui/material/IconButton";
import Menu from "@mui/material/Menu";
import MenuItem from "@mui/material/MenuItem";
import ListItemIcon from "@mui/material/ListItemIcon";
import Logout from "@mui/icons-material/Logout";
import Image from "next/image";

import { Divide as Hamburger } from "hamburger-react";
import NavbarCSS from "./Navbar.module.scss";

function Navbar() {
  const initialState = {
    openMenu: false,
    navbarMoved: false,
    is_loggedin: false,
    anchorEl: null,
    open: Boolean(null),
    navbarExpand: false,
    isAdminPanelPage: false,
  };

  function reducer(state, action) {
    switch (action.type) {
      case "user/check_admin":
        return {
          ...state,
          isAdminPanelPage: window.location.pathname.includes("adminpanel"),
        };
      case "user/check_auth":
        return {
          ...state,
          is_loggedin: action.response,
        };
      case "predicate/update":
        return {
          ...state,
          navbarExpand: window.innerWidth < 992,
        };
      case "menu/open":
        return {
          ...state,
          openMenu: !state.openMenu,
        };
      case "navbar/open":
        return {
          ...state,
          actionEl: action.target,
          open: true,
        };
      case "navbar/close":
        return {
          ...state,
          actionEl: null,
          open: false,
        };
      case "navbar/update_movement":
        return {
          ...state,
          navbarMoved: action.scrolled >= 10,
        };
      default:
        break;
    }
  }

  const [state, dispatch] = useReducer(reducer, initialState);

  function handleScroll() {
    var scrolled = document.scrollingElement.scrollTop;
    dispatch({ type: "navbar/update_movement", scrolled: scrolled });
  }
  function updatePredicate() {
    dispatch({ type: "predicate/update" });
  }

  useEffect(() => {
    document.scrollingElement.scrollTop = 0;
    updatePredicate();
    window.addEventListener("resize", updatePredicate);
    document.addEventListener("scroll", handleScroll);

    me().then((response) => {
      dispatch({ type: "user/check_auth", response: response });
    });

    return () => {
      document.removeEventListener("scroll", handleScroll);
      window.removeEventListener("resize", updatePredicate);
    };
  }, []);

  function handleLogout() {
    logout();
  }

  function handleClick(event) {
    let target = event.currentTarget;
    dispatch({ type: "navbar/open", target: target });
  }

  function handleClose() {
    dispatch({ type: "navbar/close" });
  }

  function handleGotoProfile() {
    console.log("profile profile profile profile");
    window.location.href = "http://localhost:3000/profile";
  }
  return (
    <nav
      className={
        state.isAdminPanelPage
          ? "d-none"
          : state.navbarMoved
          ? "navbar navbar-expand-lg navbar-light sticky-top nav-scrolled w-100 nav-style"
          : "navbar navbar-expand-lg navbar-light sticky-top nav-top w-100 nav-style"
      }
    >
      <div className="container-fluid">
        <a href="/" className="navbar-brand logo">
          <Image src={Logo} alt="Hotel Center" height={48} />
          <span className="fw-bold logo-text-font">Hotel Center</span>
        </a>

        {state.navbarExpand && (
          <button
            className={NavbarCSS.hamburgerButton}
            type="button"
            data-bs-toggle="collapse"
            data-bs-target="#navMenu"
          >
            <Hamburger size={20} distsnce="sm" />
          </button>
        )}

        <div className="collapse navbar-collapse" id="navMenu">
          <div className="ms-auto pt-3 pt-sm-0">
            <ul className="navbar-nav">
              {!state.is_loggedin ? (
                <Fragment>
                  {/*  {this.state.navbarExpand ? <Fragment /> : <hr />} */}
                  <hr style={{ margin: "1rem 0rem 0.5rem 0rem" }} />
                  <li className="nav-item">
                    <a href="http://localhost:3000/login">
                      <button
                        type="button"
                        className="btn btn-outline-dark me-2 nav-button nav-menu-style"
                      >
                        Login
                      </button>
                    </a>
                  </li>
                  <li className="nav-item">
                    <a href="http://localhost:3000/sign-up">
                      <button
                        type="button"
                        className="btn btn-outline-dark nav-button nav-menu-style"
                      >
                        Sign up
                      </button>
                    </a>
                  </li>
                </Fragment>
              ) : (
                <Fragment>
                  {state.navbarExpand ? (
                    <li className="nav-item">
                      <IconButton
                        onClick={handleClick}
                        size="small"
                        aria-controls={state.open ? "account-menu" : undefined}
                        aria-haspopup="true"
                        aria-expanded={state.open ? "true" : undefined}
                      >
                        <Avatar sx={{ width: 32, height: 32 }}>
                          <PersonIcon />
                        </Avatar>
                      </IconButton>
                    </li>
                  ) : (
                    <Fragment>
                      <hr />
                      <li className="nav-item">
                        <button
                          type="button"
                          className="btn btn-outline-dark nav-button nav-menu-style"
                          onClick={handleGotoProfile}
                        >
                          Profile
                        </button>
                      </li>
                      <li className="nav-item">
                        <button
                          type="button"
                          className="btn btn-outline-dark nav-button nav-menu-style"
                          onClick={() => {
                            handleLogout();
                          }}
                        >
                          Logout
                        </button>
                      </li>
                    </Fragment>
                  )}
                </Fragment>
              )}
            </ul>
          </div>
        </div>
        <Menu
          anchorEl={state.anchorEl}
          id="account-menu"
          open={state.open}
          onClose={handleClose}
          onClick={handleClose}
          PaperProps={{
            elevation: 0,
            sx: {
              overflow: "visible",
              filter: "drop-shadow(0px 2px 8px rgba(0,0,0,0.32))",
              mt: 1,
              "& .MuiAvatar-root": {
                width: 32,
                height: 32,
                ml: -0.5,
                mr: 1,
              },
              // "&:before": {
              //   content: '""',
              //   display: "block",
              //   position: "absolute",
              //   top: 0,
              //   right: 14,
              //   width: 10,
              //   height: 10,
              //   bgcolor: "background.paper",
              //   transform: "translateY(-50%) rotate(45deg)",
              //   zIndex: 0,
              // },
            },
          }}
          transformOrigin={{ horizontal: "left", vertical: "top" }}
          anchorOrigin={{ horizontal: "right", vertical: "bottom" }}
        >
          <MenuItem
            style={{ width: "150px" }}
            onClick={() => {
              handleGotoProfile();
            }}
          >
            <ListItemIcon>
              <PersonIcon fontSize="small" />
            </ListItemIcon>
            Profile
          </MenuItem>
          <MenuItem
            onClick={() => {
              handleLogout();
            }}
          >
            <ListItemIcon>
              <Logout fontSize="small" />
            </ListItemIcon>
            Logout
          </MenuItem>
        </Menu>
      </div>
    </nav>
  );
}

export default Navbar;
