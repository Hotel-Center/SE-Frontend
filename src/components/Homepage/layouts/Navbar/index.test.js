import renderer from "react-test-renderer";
import Navbar from ".";

jest.mock("next/router", () => ({
  useRouter() {
    return {
      route: "/",
      pathname: "",
      query: "",
      asPath: "",
      push: jest.fn(),
    };
  },
}));

describe("Navbar", () => {
  it("renders correctly", () => {
    const tree = renderer.create(<Navbar />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
