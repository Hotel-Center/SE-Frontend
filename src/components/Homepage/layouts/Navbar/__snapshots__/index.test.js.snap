// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar renders correctly 1`] = `
<nav
  className="navbar navbar-expand-md navbar-light sticky-top nav-top w-100 nav-style"
>
  <div
    className="container-fluid"
  >
    <a
      className="navbar-brand logo"
      href="/"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onTouchStart={[Function]}
    >
      <img
        alt="Hotel Center"
        data-nimg="1"
        decoding="async"
        height={48}
        loading="lazy"
        onError={[Function]}
        onLoad={[Function]}
        src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
        srcSet="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
        style={
          {
            "color": "transparent",
          }
        }
        width={48}
      />
      <span
        className="fw-bold logo-text-font"
      >
        Hotel Center
      </span>
    </a>
    <div
      className="collapse navbar-collapse"
      id="navMenu"
    >
      <div
        className="ms-auto pt-3 pt-sm-0"
      >
        <ul
          className="navbar-nav"
        >
          <div
            className="MuiBox-root css-11z4nni"
          >
            <a
              href="/login"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
              style={
                {
                  "marginRight": 16,
                }
              }
            >
              <button
                className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
                disabled={false}
                onBlur={[Function]}
                onContextMenu={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={0}
                type="button"
              >
                Login
              </button>
            </a>
            <a
              href="/sign-up"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              <button
                className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
                disabled={false}
                onBlur={[Function]}
                onContextMenu={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={0}
                type="button"
              >
                Sign up
              </button>
            </a>
          </div>
        </ul>
      </div>
    </div>
  </div>
</nav>
`;
